### PR TITLE
New: function-paren-newline rule (fixes #6074)

### LINKED ISSUE #6074: Rule Proposal: newline-in-function-parens
**EDIT:** I updated options in https://github.com/eslint/eslint/issues/6074#issuecomment-217124523

---

From [validateAlignedFunctionParameters](http://jscs.info/rule/validateAlignedFunctionParameters).

This will require/disallow line breaks after open parentheses of function parameters and before close parentheses of function parameters.

``` js
{
    "newline-in-function-parens": ["error", "always" or "never" or "multiline" or {"minItems": 2}]
}
```
- `"always"` (default) - Requires line breaks always.
- `"never"` - Disallows line breaks always.
- `"multiline"` - Requires line breaks when the content is multiline. Otherwise, disallows line breaks.
- `{"minItems": <integer>}` - Requires line breaks when the number of items is more than the specific integer. Otherwise, disallows line breaks.

**Valid:**

``` js
/*eslint newline-in-function-parens: ["error", "always"]*/

function foo(
) {}
function foo(
    a
) {}
function foo(
    a, b
) {}
function foo(
    a,
    b
) {}
function foo(
    a = function() {
        dosomething();
    }
) {}
```

``` js
/*eslint newline-in-function-parens: ["error", "never"]*/

function foo() {}
function foo(a) {}
function foo(a, b) {}
function foo(a,
    b) {}
function foo(a = function() {
    dosomething();
}) {}
```

``` js
/*eslint newline-in-function-parens: ["error", "multiline"]*/

function foo() {}
function foo(a) {}
function foo(a, b) {}
function foo(
    a,
    b
) {}
function foo(
    a = function() {
        dosomething();
    }
) {}
```

``` js
/*eslint newline-in-function-parens: ["error", {"minItems": 2}]*/

function foo() {}
function foo(a) {}
function foo(
    a, b
) {}
function foo(
    a,
    b
) {}
function foo(a = function() {
    dosomething();
}) {}
```

**Invalid:**

``` js
/*eslint newline-in-function-parens: ["error", "always"]*/

function foo() {}
function foo(a) {}
function foo(a, b) {}
function foo(a,
    b) {}
function foo(a = function() {
    dosomething();
}) {}
```

``` js
/*eslint newline-in-function-parens: ["error", "never"]*/

function foo(
) {}
function foo(
    a
) {}
function foo(
    a, b
) {}
function foo(
    a,
    b
) {}
function foo(
    a = function() {
        dosomething();
    }
) {}
```

``` js
/*eslint newline-in-function-parens: ["error", "multiline"]*/

function foo(
) {}
function foo(
    a
) {}
function foo(
    a, b
) {}
function foo(a,
    b) {}
function foo(a = function() {
    dosomething();
}) {}
```

``` js
/*eslint newline-in-function-parens: ["error", {"minItems": 2}]*/

function foo(
) {}
function foo(
    a
) {}
function foo(a, b) {}
function foo(a,
    b) {}
function foo(
    a = function() {
        dosomething();
    }
) {}
```

### PR DESCRIPTION
**Pull request by not-an-aardvark**
**What is the purpose of this pull request? (put an "X" next to item)**

[x] New rule (see https://github.com/eslint/eslint/issues/6074)

**What changes did you make? (Give an overview)**

This implements the `function-paren-newline` rule as described in https://github.com/eslint/eslint/issues/6074, and enables it on the ESLint codebase for dogfooding. The rule applies to all function parameters, as well as `CallExpression`/`NewExpression` arguments.

I decided to use `multiline` as the default, instead of `always` as described in https://github.com/eslint/eslint/issues/6074, because `always` seems like a very unusual style.

**Is there anything you'd like reviewers to focus on?**

* I'm unsure of the correct behavior for the `multiline` option. There are two possible behaviors that I can think of, which are subtly different:

  1. Require newlines if there are linebreaks anywhere within the parameters/arguments.
  1. Require newlines if there are any linebreaks between two parameters/arguments.

  If option 1 is used, the following code is reported, because the arguments span multiple lines. I think this is very undesirable.

  ```js
  array.map(function(x) {
    return x + 2;
  });
  ```

  If option 2 is used, the rule reports code like this, which is slightly undesirable:

  ```js
  return Boolean(
    foo && bar &&
    baz && qux
  );
  ```

  Currently, option 2 is implemented, but I'm concerned that it makes code less readable in some cases.
* The rule can produce some very long lines. For example, the following code gets reported:

  ```js
  foo(
    bar("baz", qux(
      new Foo()
    ))
  );

  // gets fixed to

  foo(bar("baz", qux(new Foo())))
  ```

  I'm concerned that this can make the code less readable.
* Due to https://github.com/eslint/eslint/issues/7522, the autofixer for this rule can be difficult to use. For example:
  ```js
  function foo() {
      foo(
          bar,
          baz);
  }

  // gets fixed to:

  function foo() {
      foo(
          bar,
          baz
  );
  }
  ```

  Since the `indent` rule doesn't report or fix the indentation of the closing `)`, it usually ends up in the wrong place. The indent rewrite fixes this issue, so maybe we should wait until 4.0.0 before releasing this rule.

### CODE REVIEW

**Comment by platinumazure** [comment]
Do we need an autofix in the first release of this rule? We could put a fixer in for 4.0 but release a usable rule in a minor release, no?

**Comment by kaicataldo** [comment]
Looks like we have some merge conflicts here. No need to rush this though since it doesn't have to be released in a major version.

File: docs/rules/function-paren-newline.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
+# enforce consistent line breaks inside function parentheses (function-paren-newline)
+
+(fixable) The `--fix` option on the [command line](../user-guide/command-line-interface#fix) automatically fixes problems reported by this rule.
```
Comment (ilyavolodin) [inline docs/rules/function-paren-newline.md]: We don't need this anymore.

**Comment by not-an-aardvark** [comment]
Does anyone have any thoughts on how to address the usability issues described in the comment at the top?

**Comment by ilyavolodin** [comment]
I think the only way to fix this usability issue is to introduce yet another option. While option 2 is reasonable, I personally prefer option 1. I also see an issue of requiring closing brace to be on it's own line (I personally prefer it to be on the same line as last argument).

**Comment by jadengore** [comment]
What is the status on this?

**Comment by not-an-aardvark** [comment]
I've been thinking about this more. What if we added a selector option to the rule to allow greater configurability? Would that solve all the use cases we need? For example, maybe people could use that to enforce parens for nested functions.

**Comment by jainanshul** [comment]
Is there an update on this issue?

**Comment by platinumazure** [comment]
Could we maybe add a `consistent` option?

* Wouldn't enforce having or not having newlines inside function parens, in general, but
* A newline is required after an opening function paren iff a newline is present before a closing function paren

I think that would be a sensible default, as well.

Also, does this apply to function calls, function declarations, or both? If both, is it worth creating rules for (or supporting options in this rule for) different behaviors for function calls vs function declarations?

**Comment by not-an-aardvark** [comment]
> Could we maybe add a `consistent` option?

I agree, adding a `consistent` option sounds like a good idea.

I think we should make `consistent` a string option like `always` and `never`, rather than adding it to the object with `minItems`. We added a similar thing to the object for [`object-curly-newline`](http://eslint.org/docs/rules/object-curly-newline), and ended up with a confusing schema where one option affects the behavior of another.

> Also, does this apply to function calls, function declarations, or both?

It applies to all "function parens" (i.e. the parentheses around the arguments in CallExpressions and NewExpressions, and the parentheses around the parameters in FunctionDeclarations, FunctionExpressions, and ArrowFunctionExpressions).

**Comment by platinumazure** [comment]
> I think we should make consistent a string option like always and never, rather than adding it to the object with minItems.

:+1:, I had that in mind as well.

> It applies to all "function parens" (i.e. the parentheses around the arguments in CallExpressions and NewExpressions, and the parentheses around the parameters in FunctionDeclarations, FunctionExpressions, and ArrowFunctionExpressions).

I'm okay with this for now, but does it not seem likely that we'll get a request to support different behavior? (But that can certainly be added later as a non-breaking change.)

**Comment by not-an-aardvark** [comment]
@eslint/eslint-team This should be ready to merge after review. Sorry about the long bikeshedding delay.

File: docs/rules/function-paren-newline.md [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+## Rule Details
+
+This rule enforces consistent line breaks inside parentheses of function parameters or arguments.
+
+### Options
+
+This rule has a single option, which can either be a string or an object.
+
+* `"always"` requires line breaks inside all function parentheses.
```
Comment (ilyavolodin) [inline docs/rules/function-paren-newline.md]: Which one of those is default (if any)? We should not it in the documentation, if there's one.

**Comment by jainanshul** [comment]
When do you guys plan to release a new eslint version with this feature?

**Comment by not-an-aardvark** [comment]
[An hour ago](https://eslint.org/blog/2017/09/eslint-v4.6.0-released)

**Comment by jainanshul** [comment]
`multiline` rule doesn't seem to like if we break an only function parameter in a new line.

```
function test(a: string) {
  console.log(a);
}

test(
  'hello',
);

```

`consistent` is what I think I need here but `consistent` allows rules like below which is not ideal either.
```
function foo(bar,
  baz) {}
```

Is there a way to allow 1 function parameter to be on a new line?

**Comment by jstephenson** [comment]
Is there a proposed solution for @jainanshul's scenario above?

**Comment by not-an-aardvark** [comment]
No. Currently, the rule is working as designed, but if you'd like to propose an enhancement to this rule, please create a new issue.

**Pull request merged**